File: engine/adapters/pi/editor.py
"""Transactional Pi v3 file editor."""
from __future__ import annotations

import json
import shutil
from pathlib import Path

from ...errors import ConcurrentModificationError, OperationUnsupportedError
from ...operations.snapshots import snapshot_file
from ..shared.codec import positive_turn, select_span
from ..shared.editing import EditBackend, EditDocument, hash_bytes, json_size, write_jsonl
from .codec import CODEC, TURN_INDEX
from .reader import _load


class PiBackend(EditBackend):
    name = "pi"
    operations = ("delete-turn", "rewrite", "replace-assistant-reply")

    def load(self, ref):
        path = Path(ref).resolve(strict=True)
        raw = path.read_bytes()
        records = [json.loads(line) for line in raw.decode().splitlines()
                   if line.strip()]
        return EditDocument(self.name, ref, path, records, hash_bytes(raw))

    def apply_ops(self, doc, ops):
        notes = []
        for op in ops:
            if op["op"] == "delete-turn":
                notes += CODEC.delete_turn(
                    doc, select_span(TURN_INDEX.turns(doc.data),
                                     positive_turn(op["turn"])))
            elif op["op"] == "rewrite":
                notes += CODEC.rewrite_message(
                    doc, op.get("locator") or op.get("uuid"), op["text"])
            else:
                raise OperationUnsupportedError("pi", op["op"])
        return notes

    def replace_reply(self, doc, turn, reply):
        return CODEC.replace_reply(doc, select_span(TURN_INDEX.turns(doc.data), turn), reply)

    def validate(self, doc):
        path = Path(doc.handle)
        if not doc.data or doc.data[0].get("type") != "session" \
                or doc.data[0].get("version") != 3:
            raise ValueError("Pi 会话缺少 v3 header")
        ids = [row.get("id") for row in doc.data[1:] if isinstance(row, dict)]
        if any(not value for value in ids) or len(ids) != len(set(ids)):
            raise ValueError("Pi entry id 无效或重复")
        known = set(ids)
        parents = {}
        calls, results = set(), set()
        for row in doc.data[1:]:
            parent = row.get("parentId")
            if parent is not None and parent not in known:
                raise ValueError("Pi parentId 指向不存在 entry")
            parents[row["id"]] = parent
            for field in ("targetId", "firstKeptEntryId", "fromId"):
                if field in row and row[field] not in known:
                    raise ValueError(f"Pi {field} 指向不存在 entry")
            message = row.get("message") or {}
            if message.get("role") == "assistant":
                required = {
                    "content", "api", "provider", "model", "usage",
                    "stopReason", "timestamp",
                }
                if not required.issubset(message):
                    raise ValueError("Pi assistant 缺少终态字段")
                calls.update(
                    part.get("id") for part in message.get("content", [])
                    if isinstance(part, dict) and part.get("type") == "toolCall"
                )
            elif message.get("role") == "toolResult":
                results.add(message.get("toolCallId"))
        if calls != results:
            raise ValueError("Pi 工具调用与结果未完整配对")
        for entry_id in known:
            seen, current = set(), entry_id
            while current is not None:
                if current in seen:
                    raise ValueError("Pi entry tree 存在环")
                seen.add(current)
                current = parents.get(current)

    def stats(self, doc):
        return {"count": len(doc.data), "size": json_size(doc.data)}

    def snapshot(self, doc, reason_code="snapshot.before_edit", extra=None):
        return snapshot_file(doc.handle, reason_code, self.name, extra)

    def restore_snapshot(self, snapshot, doc):
        shutil.copy(snapshot, doc.handle)

    def commit(self, doc):
        if hash_bytes(doc.handle.read_bytes()) != doc.revision:
            raise ConcurrentModificationError("源会话在预览后已变化，请重新预览")
        self.validate(doc)
        write_jsonl(doc.handle, doc.data)
        _load(doc.handle)
        return {"session_id": doc.data[0]["id"], "saved_as": str(doc.handle),
                "resume": f"pi --session {doc.handle}"}

File: crates/ferry-engine/src/adapters/claude/editor.rs
//! Claude 会话编辑后端：delete-turn / rewrite / replace-assistant-reply。
//!
//! 三个操作的轮次语义全部消费 [`super::codec`]，与 reader 共用同一份定义。

use std::path::{Path, PathBuf};

use serde_json::{Map, Value};

use crate::adapters::contracts::SessionEditor;
use crate::adapters::shared::codec::{
    positive_turn, select_span, NativeEditCodec, TurnIndex, TurnSpan,
};
use crate::adapters::shared::editing::{
    default_saved_revision, hash_bytes, json_size, EditDocument,
};
use crate::errors::{DomainError, DomainResult};
use crate::events::Event;

use super::codec::{CODEC, TURN_INDEX};
use super::editing as claude_edit;

/// manifest 声明的三个原生编辑操作，顺序不可改（`AgentAdapter` 会逐项比对）。
pub const OPERATIONS: &[&str] = &["delete-turn", "rewrite", "replace-assistant-reply"];

pub struct ClaudeBackend;

/// 从 `EditDocument` 取出 claude 的原生记录数组。
fn records(doc: &EditDocument) -> DomainResult<&Vec<Value>> {
    doc.data
        .downcast_ref::<Vec<Value>>()
        .ok_or_else(|| DomainError::internal("claude 编辑文档载荷类型不符"))
}

fn records_mut(doc: &mut EditDocument) -> DomainResult<&mut Vec<Value>> {
    doc.data
        .downcast_mut::<Vec<Value>>()
        .ok_or_else(|| DomainError::internal("claude 编辑文档载荷类型不符"))
}

fn handle(doc: &EditDocument) -> DomainResult<&PathBuf> {
    doc.handle
        .downcast_ref::<PathBuf>()
        .ok_or_else(|| DomainError::internal("claude 编辑文档句柄类型不符"))
}

/// 复制一份 span，避免持有对 `doc` 的不可变借用。
fn span_for(records: &[Value], selector: &Value) -> DomainResult<TurnSpan> {
    let spans = TURN_INDEX.turns(records);
    select_span(&spans, selector).cloned()
}

impl SessionEditor for ClaudeBackend {
    fn name(&self) -> &str {
        "claude"
    }

    fn operations(&self) -> &[&str] {
        OPERATIONS
    }

    fn load(&self, reference: &str) -> DomainResult<EditDocument> {
        let path = claude_edit::resolve(reference)?;
        let raw = std::fs::read(&path)
            .map_err(|error| DomainError::internal(format!("读取 claude 会话失败: {error}")))?;
        let data = claude_edit::load(&path)?;
        Ok(EditDocument::new(
            self.name(),
            reference,
            Box::new(path),
            Box::new(data),
            hash_bytes(&raw),
        ))
    }

    fn apply_ops(&self, doc: &mut EditDocument, ops: &[Value]) -> DomainResult<Vec<Event>> {
        let mut notes = Vec::new();
        for operation in ops {
            match operation.get("op").and_then(Value::as_str) {
                Some("delete-turn") => {
                    let turn = operation
                        .get("turn")
                        .cloned()
                        .ok_or_else(|| DomainError::internal("delete-turn 缺少 turn"))?;
                    let ordinal = positive_turn(&turn)?;
                    let span = span_for(records(doc)?, &Value::from(ordinal as i64))?;
                    notes.extend(CODEC.delete_turn(records_mut(doc)?, &span)?);
                }
                Some("rewrite") => {
                    let locator = operation
                        .get("locator")
                        .or_else(|| operation.get("uuid"))
                        .and_then(Value::as_str)
                        .unwrap_or_default()
                        .to_string();
                    let text = operation
                        .get("text")
                        .and_then(Value::as_str)
                        .ok_or_else(|| DomainError::internal("rewrite 缺少 text"))?
                        .to_string();
                    notes.extend(CODEC.rewrite_message(records_mut(doc)?, &locator, &text)?);
                }
                other => {
                    return Err(DomainError::operation_unsupported(
                        "claude",
                        other.unwrap_or(""),
                        None,
                    ))
                }
            }
        }
        Ok(notes)
    }

    fn replace_reply(
        &self,
        doc: &mut EditDocument,
        turn: &Value,
        reply: &Value,
    ) -> DomainResult<Vec<Event>> {
        let span = span_for(records(doc)?, turn)?;
        CODEC.replace_reply(records_mut(doc)?, &span, reply)
    }

    fn validate(&self, doc: &EditDocument) -> DomainResult<()> {
        claude_edit::check_invariants(records(doc)?)
    }

    fn stats(&self, doc: &EditDocument) -> DomainResult<Map<String, Value>> {
        let data = records(doc)?;
        let mut stats = Map::new();
        stats.insert("count".into(), Value::from(data.len() as i64));
        stats.insert(
            "size".into(),
            Value::from(json_size(&Value::Array(data.clone())) as i64),
        );
        Ok(stats)
    }

    fn snapshot(
        &self,
        doc: &EditDocument,
        reason_code: &str,
        extra: Option<&Map<String, Value>>,
    ) -> DomainResult<Option<PathBuf>> {
        Ok(Some(claude_edit::backup(
            handle(doc)?,
            reason_code,
            self.name(),
            extra,
        )?))
    }

    fn restore_snapshot(&self, snapshot: &Path, doc: &EditDocument) -> DomainResult<()> {
        std::fs::copy(snapshot, handle(doc)?)
            .map(|_| ())
            .map_err(|error| DomainError::internal(format!("claude 快照还原失败: {error}")))
    }

    fn commit(&self, doc: &mut EditDocument) -> DomainResult<Map<String, Value>> {
        let path = handle(doc)?.clone();
        let current = std::fs::read(&path)
            .map_err(|error| DomainError::internal(format!("读取 claude 会话失败: {error}")))?;
        if hash_bytes(&current) != doc.revision {
            return Err(DomainError::concurrent_modification(
                "源会话在预览后已变化，请重新预览",
            ));
        }
        let data = records(doc)?;
        claude_edit::save(&path, data)?;
        let cwd = data
            .iter()
            .filter_map(|record| record.get("cwd").and_then(Value::as_str))
            .find(|cwd| !cwd.is_empty())
            .unwrap_or(".")
            .to_string();
        let stem = path
            .file_stem()
            .map(|stem| stem.to_string_lossy().into_owned())
            .unwrap_or_default();
        let mut result = Map::new();
        result.insert("session_id".into(), Value::from(stem.as_str()));
        result.insert(
            "saved_as".into(),
            Value::from(path.to_string_lossy().into_owned()),
        );
        result.insert(
            "resume".into(),
            Value::from(format!("cd {cwd} && claude --resume {stem}")),
        );
        Ok(result)
    }

    fn saved_revision(
        &self,
        result: &Map<String, Value>,
        doc: &EditDocument,
    ) -> DomainResult<String> {
        default_saved_revision(self.name(), result, doc)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn session_file(root: &Path) -> PathBuf {
        let path = root.join("s.jsonl");
        let lines = [
            json!({"uuid": "u1", "parentUuid": null, "type": "user", "isSidechain": false,
                   "cwd": "/work", "message": {"role": "user", "content": "first"}}),
            json!({"uuid": "a1", "parentUuid": "u1", "type": "assistant", "isSidechain": false,
                   "cwd": "/work",
                   "message": {"type": "message", "role": "assistant",
                               "content": [{"type": "text", "text": "reply"}]}}),
            json!({"uuid": "u2", "parentUuid": "a1", "type": "user", "isSidechain": false,
                   "cwd": "/work", "message": {"role": "user", "content": "second"}}),
        ];
        let payload: String = lines
            .iter()
            .map(|line| format!("{}\n", serde_json::to_string(line).unwrap()))
            .collect();
        std::fs::write(&path, payload).unwrap();
        path
    }

    #[test]
    fn load_apply_validate_commit_round_trip() {
        let root = tempfile::tempdir().unwrap();
        let path = session_file(root.path());
        let backend = ClaudeBackend;
        let mut doc = backend.load(path.to_str().unwrap()).unwrap();
        assert_eq!(backend.stats(&doc).unwrap()["count"], json!(3));

        let changes = backend
            .apply_ops(&mut doc, &[json!({"op": "delete-turn", "turn": 1})])
            .unwrap();
        assert_eq!(changes[0].code, "edit.turn_deleted");
        backend.validate(&doc).unwrap();
        let result = backend.commit(&mut doc).unwrap();
        assert_eq!(result["session_id"], json!("s"));
        assert_eq!(result["resume"], json!("cd /work && claude --resume s"));
        assert_eq!(std::fs::read_to_string(&path).unwrap().lines().count(), 1);
    }

    #[test]
    fn commit_detects_concurrent_modification() {
        let root = tempfile::tempdir().unwrap();
        let path = session_file(root.path());
        let backend = ClaudeBackend;
        let mut doc = backend.load(path.to_str().unwrap()).unwrap();
        std::fs::write(&path, "{}\n").unwrap();
        let error = backend.commit(&mut doc).unwrap_err();
        assert_eq!(error.code, "session.concurrent_modification");
        assert_eq!(error.message(), "源会话在预览后已变化，请重新预览");
    }

    #[test]
    fn unknown_ops_are_rejected() {
        let root = tempfile::tempdir().unwrap();
        let path = session_file(root.path());
        let backend = ClaudeBackend;
        let mut doc = backend.load(path.to_str().unwrap()).unwrap();
        let error = backend
            .apply_ops(&mut doc, &[json!({"op": "nope"})])
            .unwrap_err();
        assert_eq!(error.code, "edit.operation_unsupported");
        assert_eq!(backend.operations(), OPERATIONS);
    }

    #[test]
    fn replace_reply_selects_by_ordinal_or_locator() {
        let root = tempfile::tempdir().unwrap();
        let path = session_file(root.path());
        let backend = ClaudeBackend;
        let mut doc = backend.load(path.to_str().unwrap()).unwrap();
        let reply = json!({"items": [{"kind": "text", "text": "new"}]});
        let changes = backend
            .replace_reply(&mut doc, &json!("u1"), &reply)
            .unwrap();
        assert_eq!(changes[0].params["turn"], json!(1));
        backend.validate(&doc).unwrap();
    }
}
